File: frontend/src/domain/shared/branded-types.ts
/**
 * Branded types for domain identifiers.
 *
 * These types provide compile-time type safety by preventing mixing of different ID types.
 * The unique symbol ensures types cannot be accidentally interchanged.
 */

declare const __brand: unique symbol;

type Brand<T, TBrand> = T & { readonly [__brand]: TBrand };

// Job domain
export type JobId = Brand<string, 'JobId'>;
export type StepId = Brand<string, 'StepId'>;

// VOD domain
export type VodId = Brand<string, 'VodId'>;
export type StreamerId = Brand<string, 'StreamerId'>;

// Audio domain
export type AudioFileId = Brand<string, 'AudioFileId'>;
export type SegmentId = Brand<string, 'SegmentId'>;

// Dataset domain
export type DatasetId = Brand<string, 'DatasetId'>;
export type EntryId = Brand<string, 'EntryId'>;

// Transcription domain
export type TranscriptId = Brand<string, 'TranscriptId'>;
export type CueId = Brand<string, 'CueId'>;

// Factory functions with validation
export const createJobId = (value: string): JobId => {
    if (!value || typeof value !== 'string') {
        throw new Error('JobId must be a non-empty string');
    }
    return value as JobId;
};

export const createVodId = (value: string): VodId => {
    if (!value || typeof value !== 'string') {
        throw new Error('VodId must be a non-empty string');
    }
    return value as VodId;
};

export const createSegmentId = (value: string): SegmentId => {
    if (!value || typeof value !== 'string') {
        throw new Error('SegmentId must be a non-empty string');
    }
    return value as SegmentId;
};

// Type guard helpers
export const isJobId = (value: unknown): value is JobId => {
    return typeof value === 'string' && value.length > 0;
};

export const isVodId = (value: unknown): value is VodId => {
    return typeof value === 'string' && value.length > 0;
};

File: backend/streamcraft/domain/shared/branded_types.py
"""Branded types for domain identifiers.

These types provide compile-time type safety by preventing mixing of different ID types.
Using NewType creates distinct types that cannot be accidentally interchanged.
"""

from typing import NewType


# Job domain
JobId = NewType("JobId", str)
StepId = NewType("StepId", str)

# VOD domain
VodId = NewType("VodId", str)
StreamerId = NewType("StreamerId", str)

# Audio domain
AudioFileId = NewType("AudioFileId", str)
SegmentId = NewType("SegmentId", str)

# Dataset domain
DatasetId = NewType("DatasetId", str)
EntryId = NewType("EntryId", str)

# Transcription domain
TranscriptId = NewType("TranscriptId", str)
CueId = NewType("CueId", str)


def create_job_id(value: str) -> JobId:
    """Create a JobId with validation."""
    if not value or not isinstance(value, str):
        raise ValueError("JobId must be a non-empty string")
    return JobId(value)


def create_step_id(value: str) -> StepId:
    """Create a StepId with validation."""
    if not value or not isinstance(value, str):
        raise ValueError("StepId must be a non-empty string")
    return StepId(value)


def create_vod_id(value: str) -> VodId:
    """Create a VodId with validation."""
    if not value or not isinstance(value, str):
        raise ValueError("VodId must be a non-empty string")
    return VodId(value)


def create_streamer_id(value: str) -> StreamerId:
    """Create a StreamerId with validation."""
    if not value or not isinstance(value, str):
        raise ValueError("StreamerId must be a non-empty string")
    return StreamerId(value)


def create_audio_file_id(value: str) -> AudioFileId:
    """Create an AudioFileId with validation."""
    if not value or not isinstance(value, str):
        raise ValueError("AudioFileId must be a non-empty string")
    return AudioFileId(value)


def create_segment_id(value: str) -> SegmentId:
    """Create a SegmentId with validation."""
    if not value or not isinstance(value, str):
        raise ValueError("SegmentId must be a non-empty string")
    return SegmentId(value)


def create_dataset_id(value: str) -> DatasetId:
    """Create a DatasetId with validation."""
    if not value or not isinstance(value, str):
        raise ValueError("DatasetId must be a non-empty string")
    return DatasetId(value)


def create_entry_id(value: str) -> EntryId:
    """Create an EntryId with validation."""
    if not value or not isinstance(value, str):
        raise ValueError("EntryId must be a non-empty string")
    return EntryId(value)


def create_transcript_id(value: str) -> TranscriptId:
    """Create a TranscriptId with validation."""
    if not value or not isinstance(value, str):
        raise ValueError("TranscriptId must be a non-empty string")
    return TranscriptId(value)


def create_cue_id(value: str) -> CueId:
    """Create a CueId with validation."""
    if not value or not isinstance(value, str):
        raise ValueError("CueId must be a non-empty string")
    return CueId(value)
